src/mlmd/EdgeCanvas.tsx

```diff
@@ -23,83 +27,102 @@ interface EdgeCanvasProps {
+                        position: 'absolute',
+                    },
+                    '&.edgeCanvasReverse': {
+                        left: 0,
+                        transform: 'translateX(-100%)',
+                    },
+                },
+            },
+        });
 
-      const {y1, y4} = lastNodePositions;
-      edgeLines.push(
-        <LineChart
-          key={`${i}-${j}`}
-          data={[
-            {x: 0, y: y1},
-            {x: 30, y: y1},
-            {x: edgeWidth - 30, y: y4},
-            {x: edgeWidth, y: y4},
-          ]}
-          areaVisible={false}
-          axisVisible={false}
-          gridVisible={false}
-          labelsVisible={false}
-          pathColor={'#BDC1C6'}
-          pathVisible={true}
-          pathWidth={1}
-          pathOpacity={1}
-          pointsVisible={false}
-          viewBoxHeight={viewHeight}
-          viewBoxWidth={edgeWidth}
-          pathSmoothing={0}
-        />
-      );
-      viewHeight += cardBodyHeight;
-      lastNodePositions[lastNode] += cardBodyHeight;
-    });
-    viewHeight += cardOffset;
-    lastNodePositions[lastNode] += cardOffset;
-  });
+        const lastNode = reverseEdges ? 'y1' : 'y4';
+        const lastNodePositions = {y1: 0, y4: 0};
 
-  const edgeCanvasClasses = classes(css.edgeCanvas, reverseEdges ? css.edgeCanvasReverse: '');
-  return  <div className={edgeCanvasClasses}>{edgeLines}</div>
-};
+        const edgeLines: JSX.Element[] = [];
+        cards.forEach((card, i) => {
+            card.elements.forEach((element, j) => {
+                if (!element.next) return;
+                const {y1, y4} = lastNodePositions;
+                edgeLines.push(
+                    <LineChart
+                        key={`${i}-${j}`}
+                        data={[
+                            {x: 0, y: y1},
+                            {x: 30, y: y1},
+                            {x: viewWidth - 30, y: y4},
+                            {x: viewWidth     , y: y4},
+                        ]}
+                        areaVisible={false}
+                        axisVisible={false}
+                        gridVisible={false}
+                        labelsVisible={false}
+                        pathColor={'#BDC1C6'}
+                        pathVisible={true}
+                        pathWidth={1}
+                        pathOpacity={1}
+                        pointsVisible={false}
+                        viewBoxHeight={viewHeight}
+                        viewBoxWidth={viewWidth}
+                        pathSmoothing={0} />
+                );
+                viewHeight += cardBodyHeight;
+                lastNodePositions[lastNode] += cardBodyHeight;
+            });
+            viewHeight += cardOffset;
+            lastNodePositions[lastNode] += cardOffset;
+        });
+
+        const edgeCanvasClasses = classes(css.edgeCanvas, reverseEdges && 'edgeCanvasReverse');
+        sleep(50).then(this.componentDidMount.bind(this))
```

Comment:
Why manually call `componentDidMount()` in `render()`? I'm not sure I get how the sleep timer is being used here.

Reply:
`componentDidMount` only runs once per mount (however we need this to lead after every state update. But also only when the physical dom nodes are available.

`componentDidMount` in terms of the react lifecycle runs after the elements are rendered onto the DOM.